src/operator/control_flow.cc

```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
   return entries;
 }
 
+struct WhileLoopParam : public dmlc::Parameter<WhileLoopParam> {
+  int num_args;
+  int num_outputs;
+  int num_out_data;
+  int max_iterations;
+  nnvm::Tuple<dim_t> cond_input_locs;
+  nnvm::Tuple<dim_t> cond_var_locs;
+  nnvm::Tuple<dim_t> func_input_locs;
+  nnvm::Tuple<dim_t> func_var_locs;
+  DMLC_DECLARE_PARAMETER(WhileLoopParam) {
+    DMLC_DECLARE_FIELD(num_args).set_lower_bound(2)
+    .describe("Number of input arguments, including cond and func as two symbol inputs.");
+    DMLC_DECLARE_FIELD(num_outputs).set_lower_bound(1)
+    .describe("The number of outputs of the subgraph, including outputs from the function body, and all loop variables.");
+    DMLC_DECLARE_FIELD(num_out_data).set_lower_bound(0)
+    .describe("The number of outputs from the function body.");
+    DMLC_DECLARE_FIELD(max_iterations).set_lower_bound(1)
+    .describe("Maximum number of iterations.");
+    DMLC_DECLARE_FIELD(cond_input_locs)
+    .describe("The locations of loop variables among the inputs.");
+  }
+};  // struct WhileLoopParam
+
+DMLC_REGISTER_PARAMETER(WhileLoopParam);
+
+class WhileLoopState: public LoopState {
+ public:
+  WhileLoopParam params;
+  Symbol cond;          // symbol of the `cond' subgraph
+  size_t n_iterations;  // the actual number of steps taken in this while loop, <= max_iterations
+  CachedOpPtr cond_op;
+
+  WhileLoopState(const WhileLoopParam &params, const Symbol &cond, const Symbol &func) : LoopState(func), params(params), cond(cond), n_iterations(0U), cond_op(LoopState::MakeSharedOp(cond)) {
+  }
+  template <typename T>
+  static void extract_by_loc(const std::vector<T> &array,
+                             const nnvm::Tuple<dim_t> input_locs,
+                             std::vector<T> *out) {
+    out->clear();
+    out->reserve(input_locs.ndim());
+    for (dim_t i : input_locs) {
+      out->push_back(array[i]);
+    }
+  }
+};
+
+static void WhileLoopComputeExCPU(const OpStatePtr& state_ptr,
+                                  const OpContext& ctx,
+                                  const std::vector<NDArray>& inputs,
+                                  const std::vector<OpReqType>& req,
+                                  const std::vector<NDArray>& outputs) {
```

Comment:
Since while loop always returns NDArrays whose size is max_iteration, we need a way to communicate with users where the valid data is in the output array.

Reply:
Marked as skipped par our discussion